internal/certification/certificates.go

```diff
@@ -0,0 +1,195 @@
+func (c *Certificates) GetClientCertificate() ([]byte, []byte) {
+	keyBytes := c.Certificates[c.ClientCertificateSecretKey][CertificateKeyKey]
+	certificateBytes := c.Certificates[c.ClientCertificateSecretKey][CertificateKey]
+
+	return keyBytes, certificateBytes
+}
+
+// Initialize initializes the Certificates object. Sets up a SharedInformer for the Secrets Resource.
+func (c *Certificates) Initialize() error {
+	logrus.Info("Certificates::Initialize")
+
+	var err error
+
+	c.Certificates = make(map[string]map[string][]byte)
+
+	informer, err := c.buildInformer()
+	if err != nil {
+		return fmt.Errorf(`error occurred building an informer: %w`, err)
+	}
+
+	c.informer = informer
+
+	err = c.initializeEventHandlers()
+	if err != nil {
+		return fmt.Errorf(`error occurred initializing event handlers: %w`, err)
+	}
+
+	return nil
+}
+
+// Run starts the SharedInformer.
+func (c *Certificates) Run() error {
+	logrus.Info("Certificates::Run")
+
+	if c.informer == nil {
+		return fmt.Errorf(`initialize must be called before Run`)
+	}
+
+	c.informer.Run(c.Context.Done())
+
+	<-c.Context.Done()
+
+	return nil
+}
+
+func (c *Certificates) buildInformer() (cache.SharedInformer, error) {
+	logrus.Debug("Certificates::buildInformer")
+
+	options := informers.WithNamespace(SecretsNamespace)
+	factory := informers.NewSharedInformerFactoryWithOptions(c.k8sClient, 0, options)
+	informer := factory.Core().V1().Secrets().Informer()
+
+	return informer, nil
+}
+
+func (c *Certificates) initializeEventHandlers() error {
+	logrus.Debug("Certificates::initializeEventHandlers")
+
+	var err error
+
+	handlers := cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.handleAddEvent,
+		DeleteFunc: c.handleDeleteEvent,
+		UpdateFunc: c.handleUpdateEvent,
+	}
+
+	c.eventHandlerRegistration, err = c.informer.AddEventHandler(handlers)
+	if err != nil {
+		return fmt.Errorf(`error occurred registering event handlers: %w`, err)
+	}
+
+	return nil
+}
+
+func (c *Certificates) handleAddEvent(obj interface{}) {
+	logrus.Debug("Certificates::handleAddEvent")
+
+	secret, ok := obj.(*corev1.Secret)
+	if !ok {
+		logrus.Errorf("Certificates::handleAddEvent: unable to cast object to Secret")
+		return
+	}
+
+	c.Certificates[secret.Name] = map[string][]byte{}
+
+	for k, v := range secret.Data {
+		c.Certificates[secret.Name][k] = v
+	}
+
+	logrus.Debugf("Certificates::handleAddEvent: certificates (%d)", len(c.Certificates))
+}
+
+func (c *Certificates) handleDeleteEvent(obj interface{}) {
+	logrus.Debug("Certificates::handleDeleteEvent")
+
+	secret, ok := obj.(*corev1.Secret)
+	if !ok {
+		logrus.Errorf("Certificates::handleDeleteEvent: unable to cast object to Secret")
+		return
+	}
+
+	if c.Certificates[secret.Name] != nil {
+		delete(c.Certificates, secret.Name)
+	}
+
+	logrus.Debugf("Certificates::handleDeleteEvent: certificates (%d)", len(c.Certificates))
+}
+
+func (c *Certificates) handleUpdateEvent(obj interface{}, obj2 interface{}) {
```

Comment:
BUG! First parameter is old state, second parameter is new state. Fix It! Rename the parameters for clarity.